components/ipfs/ipfs_utils.cc

```diff
@@ -460,47 +464,110 @@ std::string GetRegistryDomainFromIPNS(const GURL& url) {
       cid, net::registry_controlled_domains::INCLUDE_PRIVATE_REGISTRIES);
 }
 
-// gateway.io/ipfs/bafybeiemxf5abjwjbikoz4mc3a3dla6ual3jsgpdr4cjr3oz3evfy ->
-// ipfs://bafybeiemxf5abjwjbikoz4mc3a3dla6ual3jsgpdr4cjr3oz3evfy
-// bafybeiemxf5abjwjbikoz4mc3a3dla6ual3jsgpdr4cjr3oz3evfy.ipfs.gateway.io ->
-// ipfs://bafybeiemxf5abjwjbikoz4mc3a3dla6ual3jsgpdr4cjr3oz3evfy
-absl::optional<GURL> TranslateToCurrentGatewayUrl(const GURL& url) {
-  if (!url.is_valid() || !url.SchemeIsHTTPOrHTTPS()) {
-    return absl::nullopt;
+// https://github.com/ipfs/specs/blob/main/http-gateways/SUBDOMAIN_GATEWAY.md#host-request-header
+std::string DecodeSingleLabelForm(const std::string& input) {
+  if (input.find('.') != std::string::npos) {
+    return input;
   }
+  std::string result;
+  const char* chars = input.c_str();
+  size_t i = 0;
+  for (i = 0; i < input.size(); i++) {
+    if (chars[i] == '-' && (i < input.size() - 1) && chars[i + 1] == '-') {
+      result.push_back('-');
+      i++;
+    } else if (chars[i] == '-') {
+      result.push_back('.');
+    } else {
+      result.push_back(chars[i]);
+    }
+  }
+  return result;
+}
 
+absl::optional<GURL> ExtractSourceFromGatewayHost(const GURL& url) {
```

Comment:
It would be easier to read if you can have a comment to describe which cases are handled by this and which are handled by `ExtractSourceFromGatewayPath`. Also, shouldn't these two functions be moved to anonymous namespace?